src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+(ns metaprob.autodiff
+  (:refer-clojure :exclude [+ * - / == < > <= >= zero? number? pos? neg?]))
+
+;; Dual numbers represent computations dependent on a value with respect to which we are differentiating.
+;; derivative contains the current derivative; value contains the current value; and tag is a special tag
+;; that helps distinguish between multiple "levels" of nested differentiations.
+(defn make-dual-number [tag value derivative]
+  {:tag tag :value value :derivative derivative})
+
+(declare * + - /)
+
+(def non-dual? clojure.core/number?)
+
+;; Take a unary operation f on numbers
+;; and lift it to work with dual numbers.
+;; Note that df-dx should itself be a lifted
+;; function for computing the derivative of f
+;; at a point.
+(defn lift-real->real
+  [f df-dx]
+  (fn new-f [x]
+    (if (non-dual? x)
+      (f x)
+      (make-dual-number (:tag x)
```

Comment:
Could you explain a bit the reason for using `x` as the tag in this case?  (Not that I have an alternative in mind, just wondering)

Reply:
In Clojure, `(:tag x)` extracts the `:tag` field of the variable `x`. So, `x` is not a symbol but a variable holding an existing dual number, and `new-f` is returning a new dual number that uses the same tag that `x` uses. You can think of the tag as specifying a particular derivative computation; if multiple are happening at the same time (because, e.g., we are differentiating `df/dx` with respect to some variable), the tags help disambiguate.

Reply:
Oh whoops - I misread.  Got it, thanks!